4.4 Add prompting to input options

markstory opened the pull request:
When an input option is has a prompt we can now prompt for the missing value. This works for both required and optional options. A prompt value will overwrite the default if no value was provided.

Closes #15782

markstory committed:
Add prompting to input options

When an input option is has a prompt we can now prompt
for the missing value. This works for both required and optional
options. A prompt value will overwrite the default if no value was
provided.

othercorey commented:
> This works for both required and optional options.

If a prompt is set, and a user must input a value, isn't that simply a required option? What's the distinction here?

markstory commented:
> If a prompt is set, and a user must input a value, isn't that simply a required option? What's the distinction here?

I expressed what happens poorly, sorry about that. With both `prompt` and `required`, the user will be prompted for the required option when it is missing. With only `required` parsing will fail and an error will be raised.

othercorey commented:
But if an optional has a prompt isnt that just required?

saeideng commented on src/Console/ConsoleOptionParser.php:
```diff
@@ -677,18 +679,19 @@ public function subcommands()
      * @throws \Cake\Console\Exception\ConsoleException When an invalid parameter is encountered.
      */
-    public function parse(array $argv): array
+    public function parse(array $argv, ?ConsoleIo $io): array
```
optional Argument is BC, but not passing it will cause the `prompt` to be ignored

markstory replied:
Yes. We could raise an error if a prompt is used with no `ConsoleIo` so the failure isn't silent.

saeideng replied:
could we have deprecation warning if user not pass it ?
and making required Argument in cake 5?

markstory replied:
yes, I plan on making this parameter required in 5.x

saeideng commented:
another way is handling `promp` on `BaseCommand::run()` 
instead of within parser

markstory commented:
> another way is handling promp on BaseCommand::run() instead of within parser

How would that work with required options?

markstory commented:
> But if an optional has a prompt isnt that just required

The end user could provide an empty string. We don't have validation on cli options outside of using `choices`.

markstory committed:
Improve backwards compatibility and add error

Don't silently fail to prompt when using old call signatures.

othercorey commented on src/Console/ConsoleInputOption.php:
```diff
@@ -266,6 +276,34 @@ public function validChoice($value): bool
+    public function promptForInput(ConsoleIo $io): ?string
+    {
+        if (!$this->prompt) {
+            return null;
```
Shouldn't this be filtered out before calling? I'm not sure prompt should return null.

markstory replied:
If I don't have this return here, then there are type errors with `askChoice()` and `ask()`

othercorey replied:
I guess I'm suggesting that an invalid prompt shouldn't return null but should be a configuration error. Are we calling promptForInput() on all options regardless?

saeideng commented:
> How would that work with required options?

it s better to handle all related `$io` process in `Command` or   `ConsoleOptionParser` classes 

after next your changes I think the `ConsoleOptionParser` is better place to do it

I thought about this
about passing `$io` to `ConsoleInputOption::promptForInput()` 
I think this could be change

when now we have `$io` in `ConsoleOptionParser` , it could be handle here

other suggestion could be :
1) renaming `promptForInput()` to `getPrompt()`/`prompt()` or something like this 

```php
public function prompt(): ?string
{
        if (!$this->prompt) {
            return null;
        }
        return $this->prompt;
}
```
2) add `choices()` to `ConsoleInputOption`
```php
    public function choices(): ?array // only `: array` ?
    {
        return $this->_choices
```

3) handle the `ask()` / `askChoice` in `ConsoleOptionParser`

4) no need to `hasPrompt()`

markstory commented:
I can try moving all the input handling logic into `ConsoleOptionParser`

markstory committed:
Extract logic from ConsoleInputOption

Longer term I think the parse() method will need some refactoring as it
is getting quite long.

saeideng reviewed (commented):
Your last changes about moving to  `ConsoleOptionParser` looks good to me

saeideng commented:
> But if an optional has a prompt isnt that just required?

Implementing until now in this PR:
Looks : you can have `default` value , but `prompt` will ask you to set one
Am I right?

I'm not sure both these should be happen at same time

markstory commented:
> you can have default value , but prompt will ask you to set one Am I right?

Good catch, I'll get that fixed.

markstory committed:
Fix problem with default values and prompts

When you include both a prompt and a default the prompt will not ever be
used, but that seems ok as the default is used.

saeideng commented:
do we need throw `ConsoleException` in  `ConsoleInputOption`  ( within `__construct()` maybe)
when user set both `default` and `prompt` ?

markstory committed:
Throw an error when prompt and default are used together.

saeideng commented on tests/TestCase/Console/ConsoleOptionParserTest.php:
```diff
@@ -341,7 +355,119 @@ public function testAddOptionRequiredMissing(): void
+            'prompt' => 'What is your favorite?',
+        ]);
+        $out = new ConsoleOutput();
+        $io = new ConsoleIo($out, new ConsoleOutput(), new ConsoleInput(['red']));
```
looks the `'red'` is redundant IMO , since `['--color', 'blue']` provided, and `$this->assertCount(0, $out->messages());`

markstory committed:
Remove redundant input value

saeideng reviewed (approved):
Looks good to me
only `CI / Windows` has to pass

markstory committed:
Fix tests on windows.
